rfc/src/rfcs/0008-source-coverage.md

```diff
@@ -0,0 +1,277 @@
+ contains additional information (e.g., the coverage mappings) required to
+ interpret the profiling results.
+ 2. **The `.profraw` files** which essentially includes the profiling results
+ (counter and expression values) for each function of the corresponding source
+ code file.
+
+For step 3, the commands will depend on what kind of results we want.
+Most likely we will have to merge the `.profraw` files and produce a `.profdata` file as follows:
+
+```sh
+llvm-profdata merge -sparse *.profraw -o output.profdata
+```
+
+Then, we can use a command such as
+
+```sh
+llvm-cov show target/debug/binary —instr-profile=output.profdata -show-line-counts-or-regions
+```
+
+to visualize the code coverage through the terminal as in the image:
+
+![Source-based code coverage with `llvm-cov`](https://github.com/model-checking/kani/assets/73246657/4f8a973d-8977-4c0b-822d-e73ed6d223aa)
+
+or the command
+
+```sh
+llvm-cov report target/debug/binary --instr-profile=output.profdata --show-region-summary
+```
+
+to produce coverage summaries like this:
+
+```
+Filename                                             Regions    Missed Regions     Cover   Functions  Missed Functions  Executed       Lines      Missed Lines     Cover    Branches   Missed Branches     Cover
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+/long/long/path/to/my/project/binary/src/main.rs           9                 3    66.67%           3                 1    66.67%          14                 4    71.43%           0                 0         -
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+TOTAL                                                      9                 3    66.67%           3                 1    66.67%          14                 4    71.43%           0                 0         -
+```
+
+[^note-source]: The LLVM project refers to their own coverage feature as "source-based code coverage".
+It's not rare to see the term "region coverage" being used instead to refer to the same thing.
+That's because LLVM's source-based code coverage feature can report coverage for code regions,
+but other coverage frameworks don't support the concept of code regions.
+
+### The Kani coverage workflow
+
+The proposed Kani coverage workflow imitates the LLVM coverage workflow as much as possible.
+
+The two main components of the Kani coverage workflow will be the following:
+ 1. A new subcommand `cov` that drives the coverage workflow in Kani and
+    produces machine-readable coverage results.
+ 2. A new tool `kani-cov` that consumes the machine-readable coverage results
```

Comment:
It's not clear to me why we need our own tool to produce human-readable outputs.

I think I understand the need for step 1. And I agree that there needs to be a post-processing tool for some of the steps you describe below, like merging the `.kaniraw` files and excluding stdlib results.

However, why can't we make the final output (after merging) of `kani-cov` be a `.profraw` file that can be given to `llvm-cov`? This would presumably allow for the same coverage summary and terminal visualization as you have described above, for free.

If the answer is that `llvm-cov` also requires the instrumented binary (which you're replacing here with `*.kanimap`)---could we make it compatible with LLVM by writing out a degenerate LLVM file that contains only a coverage map section, and pass that to `llvm-cov`?

(I'm sure you have good answers to these but it's just not clear to me. Below, it says

> At the moment, there's only another unstable, GCC-compatible code coverage implementation based on the Gcov format. The Gcov format is line-based so it's not able to report region coverage results.

but that seems contradicted by the fact that `llvm-cov` does support region coverage and seems to give us what we need here)

Reply:
This should be much clearer now thanks to the design and future work sections. Resolving.